### ISSUE TITLE
Support Node require in ES Modules

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#29976]
## Search Terms

node require commonjs webpack conditional dynamic import

## Suggestion

Node style `require`'s are often used *inside ES Modules* for conditional imports. For example:

``` ts
const getSentry = () => {
  if (__CLIENT__) {
    return require('./client-sentry').default;
  } else {
    return require('./server-sentry').default;
  }
};

export default getSentry();
```

In this example, `require` returns `any` as per the typings provide by `@types/node`.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5f59f3dc7a0965fb1767cf13c670c633bbbbf0a8/types/node/globals.d.ts#L190-L193

In order to acquire types, we have to use `import` types:

``` ts
type ClientSentry = typeof import('./client-sentry');
type ServerSentry = typeof import('./server-sentry');

const getSentry = () => {
  if (__CLIENT__) {
    return (require('./client-sentry') as ClientSentry).default;
  } else {
    return (require('./server-sentry') as ServerSentry).default;
  }
};

export default getSentry();
```

I would like to suggest that TypeScript supports `require` calls inside of ES Modules (in Node envs). This way the workaround would not be necessary. Alternatively, perhaps TypeScript could make it possible for `@types/node` to improve the return type of `require`.

Note that `require` is desirable over dynamic imports because it is synchronous.

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by NN---** [comment]
As a workaround I just overload require function, e.g.:

```ts
declare function require(m: "jquery"): JQueryStatic;
```

**Comment by Raynos** [comment]
:+1: this would be a great fix. I would like to open a PR for this in definitely typed.

**Comment by OliverJAsh** [comment]
Interestingly, `require` does have the correct types inside `.js` files: https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#commonjs-modules-are-supported

**Comment by NN---** [comment]
It is typed to kind of `any`.

`The module support in Javascript is much more syntactically forgiving than Typescript’s module support. Most combinations of assignments and declarations are supported.`

**Comment by OliverJAsh** [comment]
> It is typed to kind of `any`.

That is correct, but it seems that in `js` files TS disregards this type information and actually provides the correct types.

**Comment by Lemmingh** [comment]
The [example](https://github.com/microsoft/TypeScript/issues/29976#issue-412098238) above seems not complete.

[`require()` does not exist in ES module](https://nodejs.org/docs/latest-v14.x/api/esm.html#esm_no_require_exports_or_module_exports) on Node.js since [draft](https://github.com/nodejs/node-eps/blob/master/002-es-modules.md#451-environment-variables). It needs to be constructed with helpers:

```javascript
import { createRequire } from "module";
const require = createRequire(import.meta.url);
```

Then, I guess changing `@types/node` type definitions won't help, but TypeScript needs to resolve semantics for `module.createRequire()`.

**Comment by andrewbranch** [comment]
Note that in #44501, the TS `import Foo = require("...")` will be desugared to a `createRequire`’d require call.
